include/substrait/function/Function.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "substrait/function/FunctionSignature.h"
+#include "substrait/type/Type.h"
+
+namespace io::substrait {
+
+struct FunctionArgument {
+  [[nodiscard]] virtual bool isRequired() const = 0;
+
+  /// Convert argument type to short type string based on
+  /// https://substrait.io/extensions/#function-signature-compound-names
+  [[nodiscard]] virtual std::string toTypeString() const = 0;
+
+  virtual bool isWildcardType() const {
+    return false;
+  };
+
+  [[nodiscard]] virtual bool isValueArgument() const {
+    return false;
+  }
+};
+
+using FunctionArgumentPtr = std::shared_ptr<FunctionArgument>;
+
+struct EnumArgument : public FunctionArgument {
+  bool required{};
+
+  [[nodiscard]] bool isRequired() const override {
+    return required;
+  }
+
+  [[nodiscard]] std::string toTypeString() const override {
+    return required ? "req" : "opt";
+  }
+};
+
+struct TypeArgument : public FunctionArgument {
+  [[nodiscard]] std::string toTypeString() const override {
+    return "type";
+  }
+
+  [[nodiscard]] bool isRequired() const override {
+    return true;
+  }
+};
+
+struct ValueArgument : public FunctionArgument {
+  ParameterizedTypePtr type;
+
+  [[nodiscard]] std::string toTypeString() const override {
+    return type->signature();
+  }
+
+  [[nodiscard]] bool isRequired() const override {
+    return true;
+  }
+
+  [[nodiscard]] bool isWildcardType() const override {
+    return type->isWildcard();
+  }
+
+  [[nodiscard]] bool isValueArgument() const override {
+    return true;
+  }
+};
+
+struct FunctionVariadic {
+  int min;
+  std::optional<int> max;
+};
+
+struct FunctionVariant {
+  std::string name;
+  std::string uri;
+  std::vector<FunctionArgumentPtr> arguments;
+  ParameterizedTypePtr returnType;
+  std::optional<FunctionVariadic> variadic;
+
+  /// Test if the actual types matched with this function variant.
+  virtual bool tryMatch(const FunctionSignature& signature);
+
+  /// Create function signature by given function name and arguments.
+  static std::string signature(
+      const std::string& name,
+      const std::vector<FunctionArgumentPtr>& arguments);
```

Comment:
Should this be an implementation detail hidden in the `cpp` file?  It could be made an anonymous namespace function.  Or at least it should be private.